crates/red_knot_python_semantic/src/semantic_index/re_exports.rs

```diff
@@ -29,39 +29,57 @@ use crate::{module_name::ModuleName, resolve_module, Db};
 #[salsa::tracked(return_ref)]
 pub(super) fn exported_names(db: &dyn Db, file: File) -> FxHashSet<Name> {
     let module = parsed_module(db.upcast(), file);
-    let mut finder = ExportFinder::new(db, file);
+    let mut exports = FxHashSet::default();
+    let mut finder = ExportFinder::new(db, file, ScopeKind::Global, &mut exports);
     finder.visit_body(module.suite());
-    finder.exports
+    exports
 }
 
-struct ExportFinder<'db> {
+struct ExportFinder<'a, 'db> {
     db: &'db dyn Db,
     file: File,
     visiting_stub_file: bool,
-    exports: FxHashSet<Name>,
+    scope_kind: ScopeKind,
+    exports: &'a mut FxHashSet<Name>,
+    global_declarations_in_scope: FxHashSet<&'db Name>,
 }
 
-impl<'db> ExportFinder<'db> {
-    fn new(db: &'db dyn Db, file: File) -> Self {
+impl<'a, 'db> ExportFinder<'a, 'db> {
+    fn new(
+        db: &'db dyn Db,
+        file: File,
+        scope_kind: ScopeKind,
+        exports: &'a mut FxHashSet<Name>,
+    ) -> Self {
         Self {
             db,
             file,
             visiting_stub_file: file.is_stub(db.upcast()),
-            exports: FxHashSet::default(),
+            scope_kind,
+            exports,
+            global_declarations_in_scope: FxHashSet::default(),
         }
     }
 
     fn possibly_add_export(&mut self, name: &Name) {
         if name.starts_with('_') {
             return;
         }
+        if self.scope_kind.is_not_global() && !self.global_declarations_in_scope.contains(name) {
```

Comment:
I don't know how hot this code path is but we could consider using hashbrown's hash table to only compute the hash key once (for `global_declarations_in_scope` and `exports`). But that's probably a premature optimization :)

Reply:
I'll leave this for now (following our in-person conversation) but happy to look into this later if it shows up in profiling etc.!